### PR TITLE
Weak runtime shape checks of matrices

### LINKED ISSUE #24149: Matrix expressions should be less strict about shape checks
I think the fundamental problem in the matrix expressions is that it validates dimensions using symbolic, structural equality. In other words, behavior like this:

```
>>> var('n m k l', integer=True)
(n, m, k, l)
>>> MatrixSymbol("A", n, m)*MatrixSymbol("B", k, l)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/matrices/expressions/matexpr.py", line 27, in __sympifyit_wrapper
    return func(a, b)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/core/decorators.py", line 106, in binary_op_wrapper
    return func(self, other)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/matrices/expressions/matexpr.py", line 127, in __mul__
    return MatMul(self, other).doit()
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/matrices/expressions/matmul.py", line 60, in __new__
    validate(*matrices)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/matrices/expressions/matmul.py", line 249, in validate
    raise ShapeError("Matrices %s and %s are not aligned"%(A, B))
sympy.matrices.common.ShapeError: Matrices A and B are not aligned
```

The problem here is that this expression is completely valid if `m` and `k` are equal. There's no reason mathematically to assume they aren't. 

We could instead allow any multiplication of any shapes unless the inner dimensions are explicitly not equal, like `(n x k) * (k + 1 x m)`. If we did this, there would be no need for GenericIdentity or GenericZeroMatrix or GenericOneMatrix. We could just use `n = Dummy('n', integer=True); Identity(n)`.

Of course, we would still want a strict validation function where inner dimensions are required to be exactly equal, but this should be something like an optional flag to `doit`, rather than something that is automatically called in the constructor. 

One problem with this approach is the assumptions system is still not very good at handling generic constraints like `Ne(k, m)`, although in principle the new assumptions machinery is capable of representing them. This isn't just for user constraints: you would need this to validate dimensions across multiple operations, e.g., `(n x n) * (m x m) * (n + 1 x n + 1)` should never validate, because it would imply that `n = m = n + 1`.

_Originally posted by @asmeurer in https://github.com/sympy/sympy/pull/24141#discussion_r998771206_

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #24149

#### Brief description of what is fixed or changed

I briefly summarize that the matrix shape check is disabled, or tested more weakly 
and left more responsibility for user to define matrix expressions in more reasonable in the first place

#### Other comments

I've also removed some logically redundant conditions for raising sympy deprecation exceptions

#### Release Notes

- matrices
  - Allowed creating of `MatAdd`, `MatMul`, `MatPow`, `Inverse` and `HadamardProduct` with more general matrix shape. For example, previously, `MatrixSymbol('A', m, n)**2` for symbolic shape `m, n` didn't work.

### CODE REVIEW

File: sympy/matrices/expressions/matpow.py [hunk 1]
```diff
@@ -15,8 +14,6 @@ def __new__(cls, base, exp, evaluate=False, **options):
         base = _sympify(base)
         if not base.is_Matrix:
             raise TypeError("MatPow base should be a matrix")
-        if not base.is_square:
-            raise NonSquareMatrixError("Power of non-square matrix %s" % base)
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/matpow.py:19]: Maybe `is_square` should use ternary logic and then this could be `if base.is_square is False`.

File: sympy/matrices/expressions/tests/test_matpow.py [hunk 2]
```diff
@@ -104,14 +103,6 @@ def test_doit_matrix():
     raises(ValueError, lambda: MatPow(X,-2).doit())
 
 
-def test_nonsquare():
-    A = MatrixSymbol('A', 2, 3)
-    B = ImmutableMatrix([[1, 2, 3], [4, 5, 6]])
-    for r in [-1, 0, 1, 2, S.Half, S.Pi, n]:
-        raises(NonSquareMatrixError, lambda: MatPow(A, r))
-        raises(NonSquareMatrixError, lambda: MatPow(B, r))
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/tests/test_matpow.py:112]: This seems like is should raise to me since we do know that the matrices are not square.

**Comment by sylee957** [comment]
I have different demo to use `Eq` and `And` evaluation for possibly detecting `m = m + 1` False in `doit`.
I have not tried using the equivalent logic old assumption check yet.

File: sympy/matrices/expressions/matadd.py [hunk 3]
```diff
@@ -52,11 +54,11 @@ def __new__(cls, *args, evaluate=False, check=None, _sympify=True):
                 "Passing check to MatAdd is deprecated and the check argument will be removed in a future version.",
                 deprecated_since_version="1.11",
                 active_deprecations_target='remove-check-argument-from-matrix-operations')
-        if check in (True, None):
+        if check:
             if not any(isinstance(i, MatrixExpr) for i in args):
                 return Add.fromiter(args)
             validate(*args)
```
Comment (sylee957) [inline sympy/matrices/expressions/matadd.py:60]: The type check logic (`MatrixExpr` vs `Expr`)is inside this, so I'd give a reminder that I may need to distinguish and pull out the type checks in the future

**Comment by sylee957** [comment]
I have changed the shape check to `Optional[bool]` type, however, I think that there are some technical difficulties about it because of the compatibility with the existing `GenericZeroMatrix, GenericOneMatrix`.
There was also minor difficulty with the type because `MatrixBase.shape` returns `int` than `Integer`.

I've also implemented a second level of shape checks that uses more stronger `Eq` than the old assumptions check. It is able to infer $M_{n, n+1}$ is not square or such, but there can be tradeoff for speed. I've implemented it only inside `doit` logic, but I'd question if it will be necessary in the future there is another argument that if makes things slower.

Also, I think that we need to undefine `GenericOneMatrix ** -1` or should not guarantee any meaningful result for it, because it complicates the implementation.

**Comment by sylee957** [comment]
The benchmark results indicates that there is slowdown for `doit` when we use more stronger logic to test out the shape.

File: sympy/matrices/expressions/matexpr.py [hunk 4]
```diff
@@ -171,8 +171,15 @@ def cols(self):
...
-        return self.rows == self.cols
+    def is_square(self) -> Optional[bool]:
+        try:
+            diff = self.rows - self.cols
+        except TypeError:
+            # XXX This error is raised because of the existance of GenericZeroMatrix, GenericIdentity
+            return None
+        if isinstance(diff, Basic):
+            return diff.is_zero
+        return diff == 0
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/matexpr.py:182]: This all seems a bit unfortunate. Ideally there would be consistent types here. I wonder if maybe the generic matrices should return some dummy symbol as well. At least for GenericIdentity it should be possible to say that `is_square` is true but maybe this is just something to be handled in those subclasses rather than here.

**Comment by oscarbenjamin** [comment]
I was expecting at least one of these to raise:
```python
In [7]: n, m = symbols('n, m')

In [8]: M = MatrixSymbol('M', m, m)

In [9]: N = MatrixSymbol('N', n, n)

In [10]: M1 = MatrixSymbol('M1', m+1, m)

In [11]: MatAdd(M, N, M1)
Out[11]: M + N + M₁

In [12]: MatAdd(M, M1, N)
Out[12]: M + M₁ + N
```
I'm not sure what would be the best way of handling these comparisons without all to all checking.

**Comment by oscarbenjamin** [comment]
> I was expecting at least one of these to raise:

To be clear I don't think it is completely necessary that it should raise. I'm just surprised that it does not because it least this equality evaluates:
```python
In [17]: Eq(m, m+1)
Out[17]: False
```
In general though I am not sure what is an efficient way of handling a situation where many symbolic expressions are expected to be equal without doing the `N^2` comparisons.

**Comment by sylee957** [comment]
> I am not sure what is an efficient way of handling a situation where many symbolic expressions are expected to be equal without doing the N^2 comparisons.

The problem is such worst time complexity is achieved every time when the input supplied by user is correct

File: sympy/matrices/expressions/hadamard.py [hunk 5]
```diff
@@ -69,15 +69,19 @@ def __new__(cls, *args, evaluate=False, check=None):
         if len(args) == 0:
             # We currently don't have a way to support one-matrices of generic dimensions:
             raise ValueError("HadamardProduct needs at least one argument")
+
+        if not all(isinstance(arg, MatrixExpr) for arg in args):
```
Comment (sylee957) [inline sympy/matrices/expressions/hadamard.py:73]: I moved the check to here because it gives better type narrowing capability

File: sympy/matrices/expressions/tests/test_hadamard.py [hunk 6]
```diff
@@ -17,7 +17,6 @@
 def test_HadamardProduct():
     assert HadamardProduct(A, B, A).shape == A.shape
 
-    raises(ShapeError, lambda: HadamardProduct(A, B.T))
```
Comment (sylee957) [inline sympy/matrices/expressions/tests/test_hadamard.py:20]: This should make sense now

File: sympy/matrices/expressions/tests/test_hadamard.py [hunk 7]
```diff
@@ -63,17 +62,13 @@ def test_hadamard():
...
-    with raises(TypeError):
-        hadamard_product()
+
+    raises(TypeError, lambda: hadamard_product())
     assert hadamard_product(A) == A
     assert isinstance(hadamard_product(A, B), HadamardProduct)
     assert hadamard_product(A, B).doit() == hadamard_product(A, B)
-    with raises(ShapeError):
-        hadamard_product(A, C)
-        hadamard_product(A, I)
```
Comment (sylee957) [inline sympy/matrices/expressions/tests/test_hadamard.py:76]: This test is removed because of redundancy

File: sympy/matrices/expressions/tests/test_inverse.py [hunk 8]
```diff
@@ -45,11 +45,7 @@ def test_inverse():
     assert isinstance(OneMatrix(n, n).I, Inverse)
 
 def test_inverse_non_invertible():
-    raises(NonSquareMatrixError, lambda: Inverse(A))
-    raises(NonSquareMatrixError, lambda: Inverse(A*B))
-    raises(NonSquareMatrixError, lambda: ZeroMatrix(n, m).I)
     raises(NonInvertibleMatrixError, lambda: ZeroMatrix(n, n).I)
-    raises(NonSquareMatrixError, lambda: OneMatrix(n, m).I)
```
Comment (sylee957) [inline sympy/matrices/expressions/tests/test_inverse.py:52]: These should make sense now

File: sympy/matrices/expressions/matexpr.py (hunk 4 above)
Comment (sylee957) [reply sympy/matrices/expressions/matexpr.py]: This is avoided by adding `is_square` check to GenericIdentity. I'm not sure about the consequences of using dummy symbols, so they are not tried.

**Comment by sylee957** [comment]
I have intermediate status report, but the benchmark indicates that I would even want to remove the shape assertion from `MatAdd, MatMul.doit` because of the performance regression of asserting symbolic matrix shape

**Comment by oscarbenjamin** [comment]
Perhaps there could be a check only for explicit Integers.

**Comment by sylee957** [comment]
I think that this should be ready for review

**Comment by sylee957** [comment]
Okay now when the square check is finally restricted to integers, the negative benchmark results are entirely gone.

File: sympy/matrices/expressions/tests/test_special.py [hunk 9]
```diff
@@ -80,7 +79,8 @@ def test_generic_identity():
     assert A != I
 
     assert I.is_Identity
-    assert I**-1 == I
+    # XXX Should we need to undefine this?
+    # assert I**-1 == I
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/tests/test_special.py:83]: Why is this changed?

**Comment by oscarbenjamin** [comment]
I had one question but otherwise this looks good to me.

**Comment by asmeurer** [comment]
It's good to see this addressed. It might still be useful to have a method that does a strict shape check, like what used to happen automatically.
